Move menu logic and VC/Passthrough/Mic checking out of MixxxMainWindow.

rryan opened the pull request:
Drops 1000+ lines from mixxx.h/mixxx.cpp -- now they're "only" 1400 lines!

I'm out for the weekend but figured I'd get this PR started. Some notes:
- I removed the re-creation of the menu bar when toggling full screen. Bug #885890 and #1076789. I'd like to see if we still need this workaround on Trusty onwards. Maybe various bugs are fixed? In either case, @Pegasus-RPG reported that menubars are still broken on KDE so maybe this isn't the right fix after all.
- I removed the partial menubar creation during bootup -- for me, QWidget::setMenuBar takes 95% of the time in initMenuBar() so not creating the full menubar doesn't seem like much of a savings to me at the code complexity cost it introduced.
- I still have some TODO(rryan)'s in WMainMenuBar that I intend to fix before merging.

rryan committed:
Move menu logic and VC/Passthrough/Mic checking out of MixxxMainWindow.

Drops 1000+ lines from mixxx.h/mixxx.cpp -- now they're "only" 1400 lines!

daschuer commented on src/mixer/basetrackplayer.h:
```diff
@@ -100,6 +107,10 @@ class BaseTrackPlayerImpl : public BaseTrackPlayer {
+    QScopedPointer<ControlObjectSlave> m_pInputEnabled;
+    QScopedPointer<ControlObjectSlave> m_pPassthroughEnabled;
+    QScopedPointer<ControlObjectSlave> m_pVinylControlEnabled;
+    QScopedPointer<ControlObjectSlave> m_pVinylControlStatus;
```
QScopedPointer here is redundant. ControlObjectSlave is deleted by the Qt object tree if it is parent to its widget.

rryan replied:
That's true, though I would like to avoid a raw pointer when possible since it documents the "ownership".

daschuer replied:
The parent pointer documents the ownership as well, as all other QObject based classes. 
A scoped pointer here will confuse the reader and requires extra CPU.
It is not obvious why we use it here and not for any other QObject based object.

rryan replied:
> The parent pointer documents the ownership as well, as all other QObject based classes.

This is true in that all code is its own documentation. But it's not very good documentation since it requires you to read the entire implementation to understand it. This documents the fact in the type of the variable which is much more clear.

> A scoped pointer here will confuse the reader 

A scoped pointer is a standard engineering practice -- there's really nothing confusing about it. 

> requires extra CPU.

Are you serious about this?

daschuer replied:
>> A scoped pointer here will confuse the reader

> A scoped pointer is a standard engineering practice -- there's really nothing confusing about it.

Yes, there is nothing against Scoped pointers to avoid delete calls. 
But in this case it is confusing, because it is redundant. The reader might think is parenting not enough? Why a scope pointer required here and not elsewhere.  
 
>>  requires extra CPU.

> Are you serious about this?

Not relay, But it is just a fact that this requires some CPU cycles without any gain.

rryan replied:
> Are you serious about this?
> 
> Not relay, But it is just a fact that this requires some CPU cycles without any gain.

QScopedPointer is very simple and does nothing more than initialize the pointer in its constructor and delete it in the destructor. The compiler will trivially inline everything so this is not true that it requires more CPU cycles.

daschuer replied:
It is not worth to discuss this in the context of i5/i7 and Mixxx, the differences are only tiny and a maintainable code is more important. But I am an original embedded developer, so I wanted to know what is correct. I compiled some test files and the result is interesting:  

In this context it is obvious, that a redundant delete is more asm code than without. 
It dependents on the implementation of the Qt object tree if it consumes more cpu cycles or not. 
I did not look at this.    

This QScopedPointer member variables are consuming much more cpu cycles, since we 
call the default constructor fist and then we call reset to adopt the final pointer. I was also surprised to see "call" statements in the asm output. This is due to the PIMPL pattern used for  QScopedPointer, instead of using an inline implementation as I have original assumed 

An interesting remaining question how compares a new/delete and QScopedPointer and std::unique_ptr in function scope. std::unique_ptr is inline implemented and the compiler is able to optimize it a lot, but it still takes a hand full  more cycles then a new/delete solution because of its custom deleter feature. Only a simple self implemented Scoped pointer produces exactly the same asm code that the original new/delete version. 

Conclusion:  
prefere std::unique_ptr over QScopedPointer

daschuer replied:
.. and these are working how we might expect it: 
boost::scoped_ptr and the deprecated std::auto_ptr

rryan replied:
Are you using -O3 and a release build (to drop the Q_ASSERTs? I did and the
assembly was identical.

On Wed, Jan 20, 2016, 3:24 AM Daniel Schürmann <notifications@github.com>
wrote:

> In src/mixer/basetrackplayer.h
> <https://github.com/mixxxdj/mixxx/pull/858#discussion_r50242825>:
>
> > @@ -100,6 +107,10 @@ class BaseTrackPlayerImpl : public BaseTrackPlayer {
> >      ControlObjectSlave* m_pPreGain;
> >      ControlObjectSlave* m_pSpeed;
> >      ControlObjectSlave* m_pPitchAdjust;
> > +    QScopedPointer<ControlObjectSlave> m_pInputEnabled;
> > +    QScopedPointer<ControlObjectSlave> m_pPassthroughEnabled;
> > +    QScopedPointer<ControlObjectSlave> m_pVinylControlEnabled;
> > +    QScopedPointer<ControlObjectSlave> m_pVinylControlStatus;
>
> .. and these are working how we might expect it:
> boost::scoped_ptr and the deprecated std::auto_ptr
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/858/files#r50242825>.
>

rryan replied:
> This QScopedPointer member variables are consuming much more cpu cycles, since we
> call the default constructor fist and then we call reset to adopt the final pointer.

I don't consider default constructing a pointer -- a single XOR instruction -- "much more" :) 

If I were to write a raw pointer as a member variable and use it as I did here, I would initialize it to null first to prevent accidental use of uninitialized memory -- so there is no difference between how I would have written it with a raw pointer or scoped pointer.

daschuer replied:
Oh .. sorry for the noise, you are right, in the optimized cased all smart pointer produce almost the same code in the
function scope. This is because the whole PIMPL implementation is inlined in the Qt case as well. 
There must have been something with my eyes that I did not notice that in the fist place. 

Other in the class scope, there are is still more code differences even if I set the pointer = NULL in the initializations list. 

We do not need to discuss about how much is much more, since the difference is always only tiny compared to the convenience and safety we have if with smart pointers. 

Please remember, that I have compared a scoped pointer solutions with a plain delete. Both are redundant here because of the QObject tree which offers the same convenience.

rryan replied:
> Please remember, that I have compared a scoped pointer solutions with a plain delete. Both are redundant here because of the QObject tree which offers the same convenience. 

Indeed, and compared to the smart pointer bits we're talking about that costs much more since it involves removing the child from the parent's children list and broadcasting a "child removed" event. :)

daschuer replied:
Did you look at the source? 
This is the children delete Loop, once the widget has had children, the 
loop will run. The currentChildBeingDeleted guard, protects us from sending a QEvent::ChildRemoved in the childs destructor and from a QList::removeAt(), which 
are called in the smart pointer case. 

```
void QObjectPrivate::deleteChildren()
{
    Q_ASSERT_X(!isDeletingChildren, "QObjectPrivate::deleteChildren()", "isDeletingChildren already set, did this function recurse?");
    isDeletingChildren = true;
    // delete children objects
    // don't use qDeleteAll as the destructor of the child might
    // delete siblings
    for (int i = 0; i < children.count(); ++i) {
        currentChildBeingDeleted = children.at(i);
        children[i] = 0;
        delete currentChildBeingDeleted;
    }
    children.clear();
    currentChildBeingDeleted = 0;
    isDeletingChildren = false;
}
```

rryan replied:
Right, I'm saying this removeAt and event are a cost of using a smart
pointer which massively eclipses the cost of the hypothetical extra
instructions the smart pointer costs -- ie this discussion is silly.

On Wed, Jan 20, 2016, 11:43 PM Daniel Schürmann <notifications@github.com>
wrote:

> In src/mixer/basetrackplayer.h
> <https://github.com/mixxxdj/mixxx/pull/858#discussion_r50368688>:
>
> > @@ -100,6 +107,10 @@ class BaseTrackPlayerImpl : public BaseTrackPlayer {
> >      ControlObjectSlave* m_pPreGain;
> >      ControlObjectSlave* m_pSpeed;
> >      ControlObjectSlave* m_pPitchAdjust;
> > +    QScopedPointer<ControlObjectSlave> m_pInputEnabled;
> > +    QScopedPointer<ControlObjectSlave> m_pPassthroughEnabled;
> > +    QScopedPointer<ControlObjectSlave> m_pVinylControlEnabled;
> > +    QScopedPointer<ControlObjectSlave> m_pVinylControlStatus;
>
> Did you look at the source?
> This is the children delete Loop, once the widget has had children, the
> loop will run. The currentChildBeingDeleted guard, protects us from
> sending a QEvent::ChildRemoved in the childs destructor and from a
> QList::removeAt(), which
> are called in the smart pointer case.
>
> void QObjectPrivate::deleteChildren()
> {
>     Q_ASSERT_X(!isDeletingChildren, "QObjectPrivate::deleteChildren()", "isDeletingChildren already set, did this function recurse?");
>     isDeletingChildren = true;
>     // delete children objects
>     // don't use qDeleteAll as the destructor of the child might
>     // delete siblings
>     for (int i = 0; i < children.count(); ++i) {
>         currentChildBeingDeleted = children.at(i);
>         children[i] = 0;
>         delete currentChildBeingDeleted;
>     }
>     children.clear();
>     currentChildBeingDeleted = 0;
>     isDeletingChildren = false;
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/858/files#r50368688>.
>

daschuer replied:
My conclusion is: 
Don't use smart pointer for QObjets with a parent. All other plain pointers should be avoided. 
So please remove them here. Thanks.

daschuer commented on src/engine/enginedeck.cpp:
```diff
@@ -34,6 +34,7 @@ EngineDeck::EngineDeck(const ChannelHandleAndGroup& handle_group,
         : EngineChannel(handle_group, defaultOrientation),
           m_pConfig(pConfig),
           m_pEngineEffectsManager(pEffectsManager ? pEffectsManager->getEngineEffectsManager() : NULL),
+          m_pInputConfigured(new ControlObject(ConfigKey(getGroup(), "input_enabled"))),
```
Input configured or input enabled? 
What is the use case for this?

rryan replied:
The previous code that I'm refactoring away needed a way to check if an input was configured for the deck -- microphone/aux have an "enabled" control that indicates this so I named it input_enabled for uniformity.

daschuer commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -335,3 +345,30 @@ void BaseTrackPlayerImpl::setupEqControls() {
 }
+
+void BaseTrackPlayerImpl::slotPassthroughEnabled(double v) {
+    bool configured = m_pInputEnabled->toBool();
```
configured != enabled

rryan replied:
Yea, it's confusing -- here passthrough is being enabled, and we need to check that the input is configured. Whether it's configured is represented by an "enabled" control (for uniformity with aux and mic).

daschuer commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -335,3 +345,30 @@ void BaseTrackPlayerImpl::setupEqControls() {
+    // Warn the user if they try to enable passthrough on a player with no
+    // configured input.
+    if (!configured && passthrough) {
+        m_pPassthroughEnabled->set(0.0);
```
This resets the passthough CO during a set call. Use connectValueChangeRequest instead.

rryan replied:
Good idea -- will do.

rryan replied:
Hm, actually the BaseTrackPlayer derivative does not own the CO so a valueChangeRequest isn't quite right here and I don't think the EngineXXX object should be emitting signals, as we should be reducing the use of signals/slots in the engine.

Since this is a refactoring I'll leave this as a TODO. The existing code in HEAD already behaves this way. A proper fix will be possible when the BaseTrackPlayer derivative owns the CO API for the engine object.

daschuer replied:
Maybe this is a hint that some code is missplaced.
maybe slotPassthroughEnabled can emmit only the signal 

and void EngineDeck::slotPassingToggle(double v) can be turned into a 
connectValueChangeRequest and do the checking.

daschuer replied:
Or we can move the CO to here and use a COS in the engine.

rryan replied:
Good suggestion, though that wouldn't work because the value change request would mean this code would never run with a value that would emit the signal. The EngineDeck would have to emit a signal that BaseTrackPlayer would receive and then rebroadcast -- and that's what I want to avoid.

rryan replied:
> Or we can move the CO to here and use a COS in the engine.

Yes, long term I think the BaseTrackPlayer derivatives should own the CO API for the engine and COs should not be in the engine at all.

Pegasus-RPG replied:
Can you add a bug/blueprint to keep track of that TODO?

rryan replied:
I'll add a TODO in the code. It's not very high priority IMO.

daschuer commented on src/mixer/basetrackplayer.h:
```diff
@@ -41,6 +42,8 @@ class BaseTrackPlayer : public BasePlayer {
+    void noVinylControlInputConfigured();
 };
 
 class BaseTrackPlayerImpl : public BaseTrackPlayer {
```
The "Base" prefix and "Impl" suffix is somehow unusual. I have not yet an overview about the object hierarchy, but it looks like we can improve the naming here to clarify it.

rryan replied:
BaseTrackPlayer is the interface, BaseTrackPlayerImpl is the implementation -- this only exists for testing -- the unit tests mock BaseTrackPlayer

daschuer replied:
Qt uses an Abstract prefix for interface classes. We should follow this. 
How about: 
"TrackPlayerAbstract" 
"TrackPlayerBase" 
This would be unambiguous to me.

rryan replied:
At the time I wrote the test suite I didn't want to change every instance of BaseTrackPlayer in the codebase to BaseTrackPlayerAbstract/Interface. I don't see a strong need to do any name changing here.

daschuer replied:
I was just confused again when I read this PR. 
With a good refactoring tool the change is an issue of a few minutes. 
But I agree, that there is no strong need to change it now.

Pegasus-RPG replied:
Please file a bug to keep track of this renaming TODO as well. (I also prefer clearer naming like Daniel suggests, though would prefer to call one of them (probably Base) just TrackPlayer (since no suffix implies that it's the base class.) )

rryan replied:
I don't think that's worthy of a bug -- cleaning it up is on my radar but not for this PR

daschuer commented on src/library/scanner/libraryscanner.h:
```diff
@@ -55,13 +55,11 @@ class LibraryScanner : public QThread {
+  public slots:
     // Call from any thread to start a scan. Does nothing if a scan is already
     // in progress.
     void scan();
```
Why is scan() a slot?

rryan replied:
So that we can connect signals to it :)

daschuer replied:
Do we? I did not find where.

rryan replied:
See mixxx.cpp line 419.

daschuer replied:
Ah, thank you. Why did it work before.

rryan replied:
It didn't -- previously the QAction was hooked to a slot in MixxxMainWindow and that slot called the LibraryScanner directly. Now WMainMenuBar is directly connected to LibraryScanner.

daschuer replied:
GitHub has tricked me: 

> 1,265 src/mixxx.cpp
>153 additions, 1,112 deletions not shown 

Yes here it is: 
https://github.com/rryan/mixxx/blob/menubar/src/mixxx.cpp#L419

daschuer commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -44,6 +45,15 @@ BaseTrackPlayerImpl::BaseTrackPlayerImpl(QObject* pParent,
 
+    m_pInputEnabled.reset(new ControlObjectSlave(group, "input_enabled", this));
+    m_pPassthroughEnabled.reset(new ControlObjectSlave(group, "passthrough", this));
+    m_pPassthroughEnabled->connectValueChanged(SLOT(slotPassthroughEnabled(double)));
```
Does this connection always work here? 
It looks like we are starting to clutter our CO interface with  "input_enabled" "passthrough" and "enabled". 
What should be used when? IMHO we should use passthough only for the vinyl control case. It this still the case?

rryan replied:
The only new CO in this PR is "input_enabled" which serves the same purpose as [MicN],enabled and [AuxN],enabled. I named it "input_enabled" since "enabled" would be less clear for the [ChannelN] namespace.

BTW passthrough is not vinyl control specific, though it is confusing that you must configure the "Vinyl Control N" input to get passthrough to work if you aren't using vinyl control.

daschuer commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -335,3 +345,30 @@ void BaseTrackPlayerImpl::setupEqControls() {
+    // Warn the user if they try to enable vinyl control on a player with no
+    // configured input.
+    if (!configured && vinylcontrol_enabled) {
+        m_pVinylControlEnabled->set(0.0);
```
The same here: connectValueChangeRequest

daschuer commented on src/mixer/microphone.cpp:
```diff
@@ -15,7 +16,23 @@ Microphone::Microphone(QObject* pParent, const QString& group, int index,
     AudioInput micInput = AudioInput(AudioPath::MICROPHONE, 0, 0, index);
     pSoundManager->registerInput(micInput, pMicrophone);
+
+    m_pInputEnabled.reset(new ControlObjectSlave(group, "enabled", this));
```
configured?

rryan replied:
This is a COS -- I'm not creating a new CO.

daschuer commented on src/mixer/microphone.cpp:
```diff
@@ -15,7 +16,23 @@ Microphone::Microphone(QObject* pParent, const QString& group, int index,
+
+    // Warn the user if they try to enable talkover on a microphone with no
+    // configured input.
+    if (!configured && talkover) {
```
connectValueChangeRequest

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -0,0 +1,692 @@
+
+QString buildWhatsThis(const QString& title, const QString& text) {
+    QString preparedTitle = title;
+    return QString("%1\n\n%2").arg(preparedTitle.replace("&", ""), text);
```
preparedTitle.remove('&')

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -0,0 +1,692 @@
+    // Delete all the old VisibilityControlConnections.
+    qDeleteAll(m_viewActionConnections);
+
+    for (auto it = m_viewActions.constBegin(); it != m_viewActions.constEnd(); ++it) {
```
can we use a range based loop here?

daschuer replied:
This is the only place where we read m_viewActions. So this should be a QList of a new struct containing
```
QActions* pAction 
ConfigKey configKey
```
This way we get around the anonymous access below and the performance will be improved.

daschuer replied:
Alternative: use QList<VisibilityControlConnection> and extend the VisibilityControlConnection API for reconnect COs.

daschuer replied:
Yes ... This will avoid the qDeleteAll(m_viewActionConnections); loop above as well.

rryan replied:
> can we use a range based loop here?

Annoyingly, range based for loops on a QMap iterate the values of the map unlike std::map. :-/

rryan replied:
> Alternative: use QList<VisibilityControlConnection> and extend the VisibilityControlConnection API for reconnect COs.

Thanks, yea that's cleaner

rryan replied:
Done -- thanks for the suggestion.

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -0,0 +1,692 @@
+                                                  tr("Space", "Menubar|View|Maximize Library"))));
+    pViewMaximizeLibrary->setStatusTip(maximizeLibraryText);
+    pViewMaximizeLibrary->setWhatsThis(buildWhatsThis(maximizeLibraryTitle, maximizeLibraryText));
+    m_viewActions[ConfigKey("[Master]", "maximize_library")] = pViewMaximizeLibrary;
```
Use m_viewActions.insert()

daschuer replied:
use QList::append()

daschuer commented:
The menu bar re-factoring is a nice idea. 
The BaseDeckPlayer is still a bit opaque to me. I think a comment that explains the Object hierarchy in the BaseDeckPlayer will help her.

daschuer commented on src/widget/wmainmenubar.cpp:
```diff
@@ -0,0 +1,692 @@
+
+namespace {
+
+QString buildWhatsThis(const QString& title, const QString& text) {
```
WhatsThis? ;-) Where will it be finally displayed?

rryan replied:
I happen to be the original author of this code I'm copying from MixxxMainWindow so I can tell you that it is used to build the string we set on the WhatsThis of our menu bar actions. :)

http://doc.qt.io/qt-4.8/qaction.html#whatsThis-prop
http://doc.qt.io/qt-4.8/qwhatsthis.html

rryan replied:
(The goal is to build a whats this without introducing any other strings that need translating, so we concatenate the title and the description of hte action)

rryan commented:
> The BaseDeckPlayer is still a bit opaque to me. I think a comment that explains the Object hierarchy in the BaseDeckPlayer will help her.

Not saying this isn't worth doing (and having better code documentation in general) but I wrote that code many years ago.. I don't see a strong need to do that in this PR :).

daschuer commented:
configured ... enabled.
I think these names grown over the years, in a way that "enabled" has become ambiguous. 
This is a refactoring PR, to clean up the code.
I think we should not introduce new COs bases on the ambiguous naming scheme. This will clutter the interface even more and will make future clean ups harder.
Since we have now an alias system, we can provide a smooth transition to an improved scheme. 

How about rename "enabled" to "configured" whenever it is a read only CO, that indicated a configured state in the preferences. Than "enabled" can be always a  r/w CO that can be used to enable things on the fly.

rryan committed:
Reconnect skin visibility controls within VisibilityControlConnection instead of in WMainMenuBar.

rryan committed:
Use QString::remove instead of replace.

rryan committed:
Rename [MicN/AuxN],enabled to input_configured.

Makes input_configured read-only and provides aliases for backwards
compatibility.

rryan commented:
> configured ... enabled.
> I think these names grown over the years, in a way that "enabled" has become ambiguous. 
> This is a refactoring PR, to clean up the code.
> I think we should not introduce new COs bases on the ambiguous naming scheme. This will clutter the interface even more and will make future clean ups harder.
> Since we have now an alias system, we can provide a smooth transition to an improved scheme.

Sounds good -- I renamed everything to input_configured.

rryan commented:
Testing on Ubuntu Trusty, the menubar bug is still there. So I'll make it match master's re-creation of the menubar

daschuer commented:
I have just tested it and it still works almost. ;-) 

If you load Shade, it enables the "Show Vinyl Control Section" if you then load then Deere, the menu is still enabled, even though Deere has not that control. 
After restarting Mixxx, the issue is gone, and the  "Show Vinyl Control Section" menu is greyed out as it should.

This is done in Master by deleting the COSs. (I can't point to to the code, because GitHub is not able to render a > 1000 line diff.)

rryan committed:
Merge branch 'master' into menubar

rryan commented:
> I have just tested it and it still works almost. ;-)
> 
> If you load Shade, it enables the "Show Vinyl Control Section" if you then load then Deere, the menu is still enabled, even though Deere has not that control. 
> After restarting Mixxx, the issue is gone, and the "Show Vinyl Control Section" menu is greyed out as it should.
> 
> This is done in Master by deleting the COSs. (I can't point to to the code, because GitHub is not able to render a > 1000 line diff.)

Huh, that's odd -- I can't reproduce this behavior. In this branch I also recreate the COSs (see VisibilityControlConnection). When I switch from Shade to Deere, "Show Vinyl Control" is greyed out as it should be, and then switching back to Shade it re-enables.

rryan committed:
Add PlayerManager numberOfDecksChanged signal.

rryan committed:
Only show Vinyl Control and load actions for enabled decks.

rryan committed:
Use QAction's triggered(bool) signal instead of toggled(bool).

Prevents signal loops since triggered is only emitted in response to
user actions while toggled is emitted when programmatically invoking
QAction::setChecked.

rryan commented:
> I still have some TODO(rryan)'s in WMainMenuBar that I intend to fix before merging.

My TODOs are fixed now.

rryan committed:
Re-create WMainMenuBar when changing in and out of fullscreen.

daschuer commented:
I can still reproduce the issue with this branch merged to master. 

* Start Mixxx
* Select Deere
* Restart Mixxx and see the "Show vinyl control section" grayed out 
* Select Shade see "Show vinyl control section" active
* Select Deere see "Show vinyl control section" still active 
--> expected "Show vinyl control section" grayed out

rryan commented:
> I can still reproduce the issue with this branch merged to master.
> 
> Start Mixxx
> Select Deere
> Restart Mixxx and see the "Show vinyl control section" grayed out
> Select Shade see "Show vinyl control section" active
> Select Deere see "Show vinyl control section" still active --> expected "Show vinyl control section" grayed out

Oh nice, I can reproduce that -- thanks.

rryan committed:
Delete visibility controls before loading the new skin.

The previous skin's visibility controls were not being deleted since the
menu bar held references to them until the new skin was done loading.

rryan commented:
> Oh nice, I can reproduce that -- thanks.

Should be fixed now.

rryan commented:
> though I still have to deal with the fullscreen menubar bug.

Just tested on my Ubuntu 14.04 machine and the menubar worked both in and out of fullscreen with the workaround in place to re-create the menubar and change its native mode to false if we're going fullscreen.

daschuer commented:
Oh ... starting Mixxx with -f seams to be broken on Ubuntu Trusty.
No launch image and Mixxx is behind the unity launch bar.  

In Mixxx 2.0.0 the Windowed launch image is shown (Precise + Cinnamon). 
I wonder if this is OK, or if the launch image should be in fullscreen too.

rryan commented:
> Oh ... starting Mixxx with -f seams to be broken on Ubuntu Trusty.
> No launch image and Mixxx is behind the unity launch bar.
> 
> In Mixxx 2.0.0 the Windowed launch image is shown (Precise + Cinnamon). 
> I wonder if this is OK, or if the launch image should be in fullscreen too.

Could you explain in more detail what isn't working as you expect?

On Ubuntu Trusty when I start mixxx from a shell with -f the launch image displays windowed while Mixxx is booting. Once the UI is ready, it goes fullscreen. The menu bar is shown in its "non-native" form as we do in master. When I go out of fullscreen, the Unity-based auto-hiding menubar returns.

I think this is the same behavior as is in master.

rryan commented:
> I wonder if this is OK, or if the launch image should be in fullscreen too.

If it's the same behavior as in 2.0.0 then I'd prefer to leave it as is since my goal here is code health.

Pegasus-RPG commented:
That is the same behavior. It would be nice for the launch/boot image to be full-screen as well, but as you said, out of scope for this PR.

Pegasus-RPG commented on src/engine/engineaux.cpp:
```diff
@@ -32,13 +40,12 @@ EngineAux::EngineAux(const ChannelHandleAndGroup& handle_group, EffectsManager*
 }
 
 EngineAux::~EngineAux() {
-    delete m_pEnabled;
```
Don't you need to delete m_pInputConfigured here too?

Pegasus-RPG replied:
Ah, no you don't, thanks to QScopedPointer. Sorry for the noise.

Pegasus-RPG commented on src/engine/enginedeck.cpp:
```diff
@@ -43,6 +44,11 @@ EngineDeck::EngineDeck(const ChannelHandleAndGroup& handle_group,
+    // Make input_configured read-only.
+    m_pInputConfigured->connectValueChangeRequest(
+        this, SLOT(slotInputConfiguredChangeRequest(double)),
+        Qt::DirectConnection);
```
My understanding of the docs says that if you find yourself forcing a slot connection type that you might be "doing it wrong." http://doc.qt.io/qt-4.8/threads-qobject.html says that with DirectConnection, "The slot is executed in the emitter's thread, which is not necessarily the receiver's thread." Is this what you want here (and elsewhere where you do this in this PR?)

rryan replied:
Yea, in engine code we largely don't do things how Qt would suggest. In this case, we're disabling changes to the CO by connecting the change request to an empty slot. If this were an AutoConnection, it would take a round trip through the main thread event loop -- when all we really want is to dump the slot into an empty function call. Direct is more efficient here.

Pegasus-RPG replied:
Okay, just making sure. Thanks for the explanation.

Pegasus-RPG commented on src/mixer/playermanager.cpp:
```diff
@@ -440,6 +446,8 @@ void PlayerManager::addMicrophoneInner() {
     QString group = groupForMicrophone(index);
     Microphone* pMicrophone = new Microphone(this, group, index, m_pSoundManager,
                                              m_pEngine, m_pEffectsManager);
+    connect(pMicrophone, SIGNAL(noMicrophoneInputConfigured()),
```
Why is this in PlayerManager? Is a Microphone considered a deck/player from its perspective?

rryan replied:
Yea, PlayerManager is now a misnomer because it manages non-players.

Pegasus-RPG commented on src/mixxx.h:
```diff
@@ -203,11 +139,12 @@ class MixxxMainWindow : public QMainWindow {
 
-    // The skin loader
+    // The skin loader.
+    // TODO(rryan): doesn't need to be a member variable
```
What should it be then?

rryan replied:
It could live on the stack when constructing the skin.

Pegasus-RPG commented:
This seems good to me. As long as the moved items still work properly (did you test them all?) I'm all for code cleanup & refactoring!

daschuer commented:
> Could you explain in more detail what isn't working as you expect?

I cannot reproduce the issue on a native installed Trusty. My test was on a VirtualBox image which I cannot access today. I think we can just ignore the issue.

rryan commented:
I'd like to get this in -- any further comments?

daschuer commented:
Just tested again and it LGTM.

rryan commented:
Thanks for the comments!

sblaisot commented:
maybe this PR is related to this bug: https://bugs.launchpad.net/mixxx/+bug/1550953
can you check @rryan ?

ferranpujolcamins commented:
Hey @rryan 
your first commit 7c69876 breaks persistence of CO appearing in the menu. You can check this by toggling the coverart CO and restarting Mixxx.

rryan commented:
Thanks @ferranpujolcamins -- fixed in https://github.com/mixxxdj/mixxx/commit/eb1ce363b9ede0c617e9f1f80918cbddd13dcecd

ferranpujolcamins commented:
Thank you for fixing it!
El dia 09/04/2016 9:40 p. m., "RJ Ryan" <notifications@github.com> va
escriure:

> Thanks @ferranpujolcamins <https://github.com/ferranpujolcamins> -- fixed
> in eb1ce36
> <https://github.com/mixxxdj/mixxx/commit/eb1ce363b9ede0c617e9f1f80918cbddd13dcecd>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/858#issuecomment-207841733>
>
